View edit delete add book photos

ishchow opened the pull request:
Closes #130 

So this was kind randomly became a big change for some reason so I'll try to explain the changes. If you want to actually review the changes and look at the code, I would recommend viewing each commit on it's own, in order so that it's easier to read

1. Change the Book model so that the photos are not stored in the book class. 
    - Storing the photos as a list made it really hard and inefficient to edit and delete the photos in Firebase since you have to update the entire list of photos when the photos were stored as an array list.
2. Add a new model for Photos.
    - In order to efficiently compute differences in List<Photo> when using the BookPhotosListAdapter, each photo has to be given a uniqueID so Android knows that any two items in the old list and the new list are the same photo (their contents could be different though).
    - Adding a new model class will allow us to eventually remove the PhotoUtils class since those functions can be moved to the model. That change can't happen in this PR since there are some changes i'd like to make to user photos before I can remove the PhotoUtils class.
3. Rename PictureController to CameraHandler, move it to a different package, and add ability to remove the PictureTakenListener.
    - I renamed the PictureController to CameraHandler and moved it to a different package since I don't think it can be considered a controller in a MVC or MVVM architecture sense. I didn't really know what else it would be called so I just called it a handler.
    - I needed to add the ability to remove the PictureTakenListener to make editing and adding book photos from a ViewEditPhotos work properly. This was due to the fact that without removing the listener after the picture was taken, you could end up adding a picture as well overwriting an existing one.
4. Update database utility classes and create new view models.
5. Add an image library called Glide to our project. Glide makes it easier and more memory efficient to handle images in Android.
6. Create ListAdapter to efficiently handle changes to a list of photos for a book retrieved from Firebase. I used this instead of a FirebaseRecyclerViewAdapter in order to conform more strictly to MVVM.
7. Add activity to perfrom CRUD (Create, Read, Update, Delete) operations on photos for a book.
8. Show first photo for the book in the AddEditBookActivity

Things I left out of this PR currently:
1. Refactor the use of photos for the User class (will make separate issue/PR for this)
2. Add ability to view book image in BookInfoActivity. 
    - Is this something we want? I can add it in but I was unsure if someone else was working in the BookInfoActivity right now so I didn't want to cause merge conflicts.

AY-Chen commented on app/src/main/java/com/example/atheneum/activities/AddEditBookActivity.java:
```diff
@@ -123,6 +127,43 @@ public void onChanged(@Nullable Book book) {
+                }
+            });
+
+            bookImage.setOnClickListener(new View.OnClickListener() {
```
This on click listener only applies in edit mode. Is there a way to move it out so it also applies when creating a new book? Right now it seems tied to the existing book because of live data.

ishchow replied:
I think the listener only applies in edit mode because of the way the database stuff is done right now. Hmmmm, this might be a tough one....

ishchow replied:
Fixed

AY-Chen commented on app/src/main/res/layout/activity_view_edit_book_photos.xml:
```diff
@@ -0,0 +1,33 @@
+
+    <include layout="@layout/content_view_edit_book_photos" />
+
+    <android.support.design.widget.FloatingActionButton
```
Really tiny thing but the plus is a weird color 
Use 
`        android:tint="@android:color/white"`
to make it white

ishchow replied:
Fixed

AY-Chen commented on app/src/main/java/com/example/atheneum/activities/ViewEditBookPhotos.java:
```diff
@@ -0,0 +1,157 @@
+                            .setPositiveButton(getString(R.string.dialog_delete), new DialogInterface.OnClickListener() {
+                                @Override
+                                public void onClick(DialogInterface dialog, int which) {
+                                    bookPhotosViewModel.deletePhoto(photo);
```
Photo is not removed when returning back to the edit activity. It seems it's not being updated. 
It does change once you exit and re-enter the edit page..

ishchow replied:
I think it's because in the AddEdiBooktActivity, I do a null check when I observe for changes to the first book photo. I don't update the image view when it's null but null is returned when the data doesn't exist. That should be an easy fix.

ishchow replied:
Fixed using Glide. Glide is really cool, if you give it a fallback picture and the bitmap you try to load is null, it just loads the fallback picture.

AY-Chen commented on app/src/main/java/com/example/atheneum/activities/ViewEditBookPhotos.java:
```diff
@@ -0,0 +1,157 @@
+        if (FirebaseAuthUtils.getCurrentUser().getUid().equals(ownerUserID)) {
+            cameraHandler = CameraHandler.newInstance(this);
+
+            bookPhotosListAdapter.setOnItemClickListener(new BookPhotosListAdapter.OnItemClickListener() {
```
Any particular reason  you decided to have click on the photo open the camera? Just kind of curious because i thought it was a bit strange.

ishchow replied:
Clicking on the photo lets you edit it. When you take the picture, the photo updates in the Recyclerview. When you cancel, nothing happens. That's probably not the best UI since it's really confusing.

I could probably change it so that there is a dialog box that lets you edit/delete the photo. 

It might be better to even have a separate activity for viewing/editing/deleting a particular photo for a book. But that might depend on how good having a scaled up image of the book looks like since the images are really low res.

ishchow replied:
Fixed. This was a really bad UI design. So I just changed it so that clicking a photo takes you to a dialog box that would ask you to edit the photo while a long click takes you to a dialog box asking you to delete a photo.

ishchow commented:
> 
> 
> so the intent was to decouple photos from books? seems good to me
> 
> though for commit 6 what did you mean by FirebaseRecyclerViewAdapter not conforming to MVVM?

When you instantiate FirebaseRecyclerViewAdapter, you have to give the adapter a reference to the data to pull from Firebase. This sort of breaks the principles of MVVM since you are now leaking knowledge of the database to the activity. The idea in MVVM is to keep the view (aka activity) as dumb as possible and to know nothing about the database. The view can only interact with the database through the view model. Technically speaking. I used the BookPhotosViewModel + BookPhotosListAdapter to do the same thing as FirebaseRecyclerViewAdapter while being more architecturally correct. There are some cases where you can't do this but in this case I could replace the use of FirebaseRecyclerViewAdapter so I did.

ishchow commented:
I fixed all the things that @AY-Chen mentioned (hopefully). Also, I noticed that there was no app bar for BookInfoActivity or AddEditBookActivity so I added them back in. Since the book photos and the book are decoupled in the database, to start the ViewEditBookPhotosActivity to edit the book photos, you need to click the option "Edit Book Photos" in the toolbar in BookInfoActivity. The ViewEditBookPhotosActivity is still spawned in the AddEditBookActivity when a new book is added though just not for edit. So now the book and the photos for a book are pretty decoupled in the UI as well. Lastly, I removed the "Edit" and "Delete" buttons in BookInfoActivity and just added them to the toolbar. Trying to delete a book opens a dialog box to warn the user about this action.

ishchow commented:
@AY-Chen @MarcusBoay Can I get both of you to look at this? I had merge conflicts with both of your PRs so I just want to make sure I didn't break something accidently. I tested the goodreads stuff and the new request activity and it seems to be fine but I just want to make sure.

MarcusBoay commented:
Notifications still work as intended

ishchow commented:
Goodreads stuff seems to work and adding/editing/deleting/viewing pictures works as well. I'm going to merge this before I start getting more conflicts.
